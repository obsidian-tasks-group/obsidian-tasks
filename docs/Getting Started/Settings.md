---
publish: true
---

# Settings

## Restart after changing settings

Regrettably, many of the settings for customising Tasks require one of the following to be done, for the new options to take effect:

- Obsidian to be restarted,
- or the vault to be closed and re-opened.

The Tasks Settings pane warns of this:

![Changing any settings requires a restart of obsidian](../images/settings-restart-after-change.png)

## Available settings

As the number of options in Tasks has grown, we have documented each setting in the page for the relevant topic.

For convenience, here is a list of all those documentation pages (in the order the options appear in the Tasks Settings pane):

- [[About Task Formats#Selecting the task format]]
- [[Global Filter]]
- [[Global Query]]
- [[Status Settings]]
- [[Dates#Date-tracking settings]]
- [[Recurring Tasks#Order of the new task]]
- [[Auto-Suggest]]
- [[Create or edit Task#Turning off keyboard shortcuts]]
