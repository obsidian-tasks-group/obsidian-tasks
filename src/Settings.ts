export interface Settings {
    globalFilter: string;
    removeGlobalFilter: boolean;
    dateFormat: string;
    dateLink: boolean;
}

const defaultSettings: Settings = {
    globalFilter: '',
    removeGlobalFilter: false,
    dateFormat: 'YYYY-MM-DD',
    dateLink: false,
};

let settings: Settings = { ...defaultSettings };

export const getSettings = (): Settings => {
    return { ...settings };
};

export const updateSettings = (newSettings: Partial<Settings>): Settings => {
    settings = { ...settings, ...newSettings };

    return getSettings();
};
