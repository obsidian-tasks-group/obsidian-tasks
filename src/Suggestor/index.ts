import type { Task } from 'Task/Task';
import type { Settings } from '../Config/Settings';

/*
 * A suggestion presented to the user and some metadata about it.
 */
export type SuggestInfo = {
    suggestionType?: 'match' | 'default' | 'empty';
    /** What to display to the user  */
    displayText: string;
    /** What to append to the note  */
    appendText: string;
    /** At what index in the line to do the insertion (if not specified, the cursor location is used) */
    insertAt?: number;
    /** How many characters to skip from the original line (e.g. if replacing existing text) */
    insertSkip?: number;
    /** Text to match with the user input if matching against the display text is not desirable */
    textToMatch?: string;
    /** Task which needs too be Updated on Select */
    taskItDependsOn?: Task;
};

/*
 * Return a list of suggestions, either generic or more fine-grained to the words at the cursor.
 */
export type SuggestionBuilder = (
    line: string,
    cursorPos: number,
    settings: Settings,
    allTasks: Task[],
    taskToSuggestFor?: Task,
) => SuggestInfo[];
