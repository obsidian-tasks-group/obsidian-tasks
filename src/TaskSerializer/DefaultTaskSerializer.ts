import type { Moment } from 'moment';
import { TaskLayout } from '../TaskLayout';
import type { TaskLayoutComponent } from '../TaskLayout';
import { Recurrence } from '../Recurrence';
import { Priority, Task, TaskRegularExpressions } from '../Task';
import type { TaskDetails, TaskSerializer } from '.';

/* Interface describing the symbols that {@link DefaultTaskSerializer}
 * uses to serialize and deserialize tasks.
 *
 * @export
 * @interface DefaultTaskSerializerSymbols
 */
export interface DefaultTaskSerializerSymbols {
    readonly prioritySymbols: {
        Highest: string;
        High: string;
        Medium: string;
        Low: string;
        Lowest: string;
        None: string;
    };
    readonly startDateSymbol: string;
    readonly createdDateSymbol: string;
    readonly scheduledDateSymbol: string;
    readonly dueDateSymbol: string;
    readonly doneDateSymbol: string;
    readonly recurrenceSymbol: string;
    readonly TaskFormatRegularExpressions: {
        priorityRegex: RegExp;
        startDateRegex: RegExp;
        createdDateRegex: RegExp;
        scheduledDateRegex: RegExp;
        dueDateRegex: RegExp;
        doneDateRegex: RegExp;
        recurrenceRegex: RegExp;
    };
}

/**
 * A symbol map for obsidian-task's default task style.
 * Uses emojis to concisely convey meaning
 */
export const DEFAULT_SYMBOLS: DefaultTaskSerializerSymbols = {
    prioritySymbols: {
        Highest: 'üî∫',
        High: '‚è´',
        Medium: 'üîº',
        Low: 'üîΩ',
        Lowest: '‚è¨',
        None: '',
    },
    startDateSymbol: 'üõ´',
    createdDateSymbol: '‚ûï',
    scheduledDateSymbol: '‚è≥',
    dueDateSymbol: 'üìÖ',
    doneDateSymbol: '‚úÖ',
    recurrenceSymbol: 'üîÅ',
    TaskFormatRegularExpressions: {
        // The following regex's end with `$` because they will be matched and
        // removed from the end until none are left.
        priorityRegex: /([üî∫‚è´üîºüîΩ‚è¨])$/u,
        startDateRegex: /üõ´ *(\d{4}-\d{2}-\d{2})$/u,
        createdDateRegex: /‚ûï *(\d{4}-\d{2}-\d{2})$/u,
        scheduledDateRegex: /[‚è≥‚åõ] *(\d{4}-\d{2}-\d{2})$/u,
        dueDateRegex: /[üìÖüìÜüóì] *(\d{4}-\d{2}-\d{2})$/u,
        doneDateRegex: /‚úÖ *(\d{4}-\d{2}-\d{2})$/u,
        recurrenceRegex: /üîÅ ?([a-zA-Z0-9, !]+)$/iu,
    },
} as const;

export class DefaultTaskSerializer implements TaskSerializer {
    constructor(public readonly symbols: DefaultTaskSerializerSymbols) {}

    /* Convert a task to its string representation
     *
     * @param task The task to serialize
     *
     * @return The string representation of the task
     */
    public serialize(task: Task): string {
        const taskLayout = new TaskLayout();
        let taskString = '';
        for (const component of taskLayout.layoutComponents) {
            taskString += this.componentToString(task, taskLayout, component);
        }
        return taskString;
    }

    /**
     * Renders a specific TaskLayoutComponent of the task (its description, priority, etc) as a string.
     */
    public componentToString(task: Task, layout: TaskLayout, component: TaskLayoutComponent) {
        const {
            prioritySymbols,
            startDateSymbol,
            createdDateSymbol,
            scheduledDateSymbol,
            doneDateSymbol,
            recurrenceSymbol,
            dueDateSymbol,
        } = this.symbols;

        switch (component) {
            case 'description':
                return task.description;
            case 'priority': {
                let priority: string = '';

                if (task.priority === Priority.Highest) {
                    priority = ' ' + prioritySymbols.Highest;
                } else if (task.priority === Priority.High) {
                    priority = ' ' + prioritySymbols.High;
                } else if (task.priority === Priority.Medium) {
                    priority = ' ' + prioritySymbols.Medium;
                } else if (task.priority === Priority.Low) {
                    priority = ' ' + prioritySymbols.Low;
                } else if (task.priority === Priority.Lowest) {
                    priority = ' ' + prioritySymbols.Lowest;
                }
                return priority;
            }
            case 'startDate':
                if (!task.startDate) return '';
                return layout.options.shortMode
                    ? ' ' + startDateSymbol
                    : ` ${startDateSymbol} ${task.startDate.format(TaskRegularExpressions.dateFormat)}`;
            case 'createdDate':
                if (!task.createdDate) return '';
                return layout.options.shortMode
                    ? ' ' + createdDateSymbol
                    : ` ${createdDateSymbol} ${task.createdDate.format(TaskRegularExpressions.dateFormat)}`;
            case 'scheduledDate':
                if (!task.scheduledDate || task.scheduledDateIsInferred) return '';
                return layout.options.shortMode
                    ? ' ' + scheduledDateSymbol
                    : ` ${scheduledDateSymbol} ${task.scheduledDate.format(TaskRegularExpressions.dateFormat)}`;
            case 'doneDate':
                if (!task.doneDate) return '';
                return layout.options.shortMode
                    ? ' ' + doneDateSymbol
                    : ` ${doneDateSymbol} ${task.doneDate.format(TaskRegularExpressions.dateFormat)}`;
            case 'dueDate':
                if (!task.dueDate) return '';
                return layout.options.shortMode
                    ? ' ' + dueDateSymbol
                    : ` ${dueDateSymbol} ${task.dueDate.format(TaskRegularExpressions.dateFormat)}`;
            case 'recurrenceRule':
                if (!task.recurrence) return '';
                return layout.options.shortMode
                    ? ' ' + recurrenceSymbol
                    : ` ${recurrenceSymbol} ${task.recurrence.toText()}`;
            case 'blockLink':
                return task.blockLink ?? '';
            default:
                throw new Error(`Don't know how to render task component of type '${component}'`);
        }
    }

    /**
     * Given the string captured in the first capture group of
     *    {@link DefaultTaskSerializerSymbols.TaskFormatRegularExpressions.priorityRegex},
     *    returns the corresponding Priority level.
     *
     * @param p String captured by priorityRegex
     * @returns Corresponding priority if parsing was successful, otherwise {@link Priority.None}
     */
    protected parsePriority(p: string): Priority {
        const { prioritySymbols } = this.symbols;
        switch (p) {
            case prioritySymbols.Lowest:
                return Priority.Lowest;
            case prioritySymbols.Low:
                return Priority.Low;
            case prioritySymbols.Medium:
                return Priority.Medium;
            case prioritySymbols.High:
                return Priority.High;
            case prioritySymbols.Highest:
                return Priority.Highest;
            default:
                return Priority.None;
        }
    }

    /* Parse TaskDetails from the textual description of a {@link Task}
     *
     * @param line The string to parse
     *
     * @return {TaskDetails}
     */
    public deserialize(line: string): TaskDetails {
        const { TaskFormatRegularExpressions } = this.symbols;

        // Keep matching and removing special strings from the end of the
        // description in any order. The loop should only run once if the
        // strings are in the expected order after the description.
        let matched: boolean;
        let priority: Priority = Priority.None;
        let startDate: Moment | null = null;
        let scheduledDate: Moment | null = null;
        let dueDate: Moment | null = null;
        let doneDate: Moment | null = null;
        let createdDate: Moment | null = null;
        let recurrenceRule: string = '';
        let recurrence: Recurrence | null = null;
        // Tags that are removed from the end while parsing, but we want to add them back for being part of the description.
        // In the original task description they are possibly mixed with other components
        // (e.g. #tag1 <due date> #tag2), they do not have to all trail all task components,
        // but eventually we want to paste them back to the task description at the end
        let trailingTags = '';
        // Add a "max runs" failsafe to never end in an endless loop:
        const maxRuns = 20;
        let runs = 0;
        do {
            matched = false;
            const priorityMatch = line.match(TaskFormatRegularExpressions.priorityRegex);
            if (priorityMatch !== null) {
                priority = this.parsePriority(priorityMatch[1]);
                line = line.replace(TaskFormatRegularExpressions.priorityRegex, '').trim();
                matched = true;
            }

            const doneDateMatch = line.match(TaskFormatRegularExpressions.doneDateRegex);
            if (doneDateMatch !== null) {
                doneDate = window.moment(doneDateMatch[1], TaskRegularExpressions.dateFormat);
                line = line.replace(TaskFormatRegularExpressions.doneDateRegex, '').trim();
                matched = true;
            }

            const dueDateMatch = line.match(TaskFormatRegularExpressions.dueDateRegex);
            if (dueDateMatch !== null) {
                dueDate = window.moment(dueDateMatch[1], TaskRegularExpressions.dateFormat);
                line = line.replace(TaskFormatRegularExpressions.dueDateRegex, '').trim();
                matched = true;
            }

            const scheduledDateMatch = line.match(TaskFormatRegularExpressions.scheduledDateRegex);
            if (scheduledDateMatch !== null) {
                scheduledDate = window.moment(scheduledDateMatch[1], TaskRegularExpressions.dateFormat);
                line = line.replace(TaskFormatRegularExpressions.scheduledDateRegex, '').trim();
                matched = true;
            }

            const startDateMatch = line.match(TaskFormatRegularExpressions.startDateRegex);
            if (startDateMatch !== null) {
                startDate = window.moment(startDateMatch[1], TaskRegularExpressions.dateFormat);
                line = line.replace(TaskFormatRegularExpressions.startDateRegex, '').trim();
                matched = true;
            }

            const createdDateMatch = line.match(TaskFormatRegularExpressions.createdDateRegex);
            if (createdDateMatch !== null) {
                createdDate = window.moment(createdDateMatch[1], TaskRegularExpressions.dateFormat);
                line = line.replace(TaskFormatRegularExpressions.createdDateRegex, '').trim();
                matched = true;
            }

            const recurrenceMatch = line.match(TaskFormatRegularExpressions.recurrenceRegex);
            if (recurrenceMatch !== null) {
                // Save the recurrence rule, but *do not parse it yet*.
                // Creating the Recurrence object requires a reference date (e.g. a due date),
                // and it might appear in the next (earlier in the line) tokens to parse
                recurrenceRule = recurrenceMatch[1].trim();
                line = line.replace(TaskFormatRegularExpressions.recurrenceRegex, '').trim();
                matched = true;
            }

            // Match tags from the end to allow users to mix the various task components with
            // tags. These tags will be added back to the description below
            const tagsMatch = line.match(TaskRegularExpressions.hashTagsFromEnd);
            if (tagsMatch != null) {
                line = line.replace(TaskRegularExpressions.hashTagsFromEnd, '').trim();
                matched = true;
                const tagName = tagsMatch[0].trim();
                // Adding to the left because the matching is done right-to-left
                trailingTags = trailingTags.length > 0 ? [tagName, trailingTags].join(' ') : tagName;
            }

            runs++;
        } while (matched && runs <= maxRuns);

        // Now that we have all the task details, parse the recurrence rule if we found any
        if (recurrenceRule.length > 0) {
            recurrence = Recurrence.fromText({
                recurrenceRuleText: recurrenceRule,
                startDate,
                scheduledDate,
                dueDate,
            });
        }
        // Add back any trailing tags to the description. We removed them so we can parse the rest of the
        // components but now we want them back.
        // The goal is for a task of them form 'Do something #tag1 (due) tomorrow #tag2 (start) today'
        // to actually have the description 'Do something #tag1 #tag2'
        if (trailingTags.length > 0) line += ' ' + trailingTags;

        return {
            description: line,
            priority,
            startDate,
            createdDate,
            scheduledDate,
            dueDate,
            doneDate,
            recurrence,
            tags: Task.extractHashtags(line),
        };
    }
}
