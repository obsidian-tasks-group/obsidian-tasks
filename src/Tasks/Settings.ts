export class Settings {
    public static readonly DEFAULT_TAG = '#task';
    private static taskTag: string = Settings.DEFAULT_TAG;
    private static regexTask: RegExp = Settings.createTaskRegex(
        Settings.DEFAULT_TAG,
    );

    public static readonly DEFAULT_SETTINGS = {
        taskTag: Settings.DEFAULT_TAG,
    };

    public static readonly IDENTIFIER_DUE_DATE = 'ðŸ“…';
    public static readonly IDENTIFIER_DONE_DATE = 'âœ…';
    public static readonly DATE_FORMAT = 'YYYY-MM-DD';

    public static readonly CLASS_ITEM = 'tasks-item';

    public static readonly DATA_PAGE_INDEX = 'data-tasks-page-index';
    public static readonly DATA_PATH = 'data-tasks-path';

    public static readonly REGEX_LI_TASK = /^ ([^ðŸ“…âœ…]*)(.*)/u;
    public static readonly REGEX_DUE_DATE = /ðŸ“… (\d{4}-\d{2}-\d{2})/u;
    public static readonly REGEX_DONE_DATE = /âœ… (\d{4}-\d{2}-\d{2})/u;

    constructor({ taskTag }: { taskTag: string }) {
        Settings.taskTag = taskTag;
        Settings.regexTask = Settings.createTaskRegex(taskTag);
    }

    public static get TASK_TAG(): string {
        return Settings.taskTag;
    }

    public static get REGEX_TASK(): RegExp {
        return Settings.regexTask;
    }

    private static createTaskRegex(taskTag: string): RegExp {
        return new RegExp(
            `^([\\s\\t]*)- \\[( |x)\\] ${taskTag} ([^ðŸ“…âœ…]*)(.*)`,
            'u',
        );
    }
}
