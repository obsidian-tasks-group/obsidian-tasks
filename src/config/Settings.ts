import { Feature } from './Feature';
import type { FeatureFlag } from './Feature';

export interface Settings {
    globalFilter: string;
    removeGlobalFilter: boolean;
    setDoneDate: boolean;
    autoSuggestInEditor: boolean;
    autoSuggestMinMatch: number;
    autoSuggestMaxItems: number;

    // Collection of feature flag IDs and their state.
    features: FeatureFlag;
}

const defaultSettings: Settings = {
    globalFilter: '',
    removeGlobalFilter: false,
    setDoneDate: true,
    autoSuggestInEditor: true,
    autoSuggestMinMatch: 0,
    autoSuggestMaxItems: 6,
    features: Feature.settingsFlags,
};

let settings: Settings = { ...defaultSettings };

/**
 * Returns the current settings as a object, it will also check and
 * update the flags to make sure they are all shown in the data.json
 * file. Exposure via the settings UI is optional.
 *
 * @export
 * @returns true if the feature is enabled.
 */
export const getSettings = (): Settings => {
    // Check to see if there is a new flag and if so add it to the users settings.
    for (const flag in Feature.settingsFlags) {
        if (settings.features[flag] === undefined) {
            settings.features[flag] = Feature.settingsFlags[flag];
        }
    }

    return { ...settings };
};

export const updateSettings = (newSettings: Partial<Settings>): Settings => {
    settings = { ...settings, ...newSettings };

    return getSettings();
};

export const resetSettings = (): Settings => {
    return updateSettings(defaultSettings);
};

/**
 * Returns the enabled state of the feature from settings.
 *
 * @export
 * @param internalName the internal name of the feature.
 * @returns true if the feature is enabled.
 */
export const isFeatureEnabled = (internalName: string): boolean => {
    return settings.features[internalName] ?? false;
};

/**
 * enables toggling the feature and returning the current collection with state.
 *
 * @export
 * @param internalName the internal name of the feature.
 * @param enabled the expected state of the feature.
 * @returns the features with the specified feature toggled.
 */
export const toggleFeature = (
    internalName: string,
    enabled: boolean,
): FeatureFlag => {
    settings.features[internalName] = enabled;
    return settings.features;
};
