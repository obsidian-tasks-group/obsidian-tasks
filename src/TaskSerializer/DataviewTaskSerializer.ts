import type { TaskLayout, TaskLayoutComponent } from '../TaskLayout';
import type { Task } from '../Task';
import { Priority } from '../Task';
import { DefaultTaskSerializer } from './DefaultTaskSerializer';

/**
 * Takes a regex of the form 'key:: value' and turns it into a regex that can parse
 * Dataview inline field, i.e either;
 *     * (key:: value)
 *     * [key:: value]
 *
 * There can be an arbitrary amount of horizontal whitespace around the key value pair,
 * and after the '::'
 */
function toInlineFieldRegex(innerFieldRegex: RegExp): RegExp {
    /**
     * First, I'm sorry this looks so bad. Javascript's regex engine lacks some
     * conveniences from other engines like PCRE (duplicate named groups)
     * that would've made this easier to express in a readable way.
     *
     * The idea here is that we're trying to say, in English:
     *
     *     "{@link innerFieldRegex} can either be surrounded by square brackets `[]`
     *     or parens `()`"
     *
     * But there is added complexity because we want to disallow mismatched pairs
     *   (i.e. no `[key::value) or (key::value]`). And we have to take care to not
     * introduce new capture groups, since innerFieldRegex may contain capture groups
     * and depend on the numbering.
     *
     * We achieve this by using a variable length, positive lookahead to assert
     * "Only match a the first element of the pair if the other element is somewhere further in the string".
     *
     * This is likely somewhat fragile.
     *
     */
    const fieldRegex = (
        [
            '(?:',
            /*     */ /(?=[^\]]+\])\[/, // Try to match '[' if there's a ']' later in the string
            /*    */ '|',
            /*     */ /(?=[^)]+\))\(/, // Otherwise, match '(' if there's a ')' later in the string
            ')',
            / */,
            innerFieldRegex,
            / */,
            /[)\]]/,
            /(?: *,)?/, // Allow trailing comma, enables workaround from #1913 for rendering issue
            /$/, // Regexes are matched from the end of the string forwards
        ] as const
    )
        .map((val) => (val instanceof RegExp ? val.source : val))
        .join('');
    return new RegExp(fieldRegex, innerFieldRegex.flags);
}

/**
 * A symbol map that corresponds to a task format that strives to be compatible with
 *   [Dataview]{@link https://github.com/blacksmithgu/obsidian-dataview}
 */
export const DATAVIEW_SYMBOLS = {
    prioritySymbols: {
        Highest: 'priority:: highest',
        High: 'priority:: high',
        Medium: 'priority:: medium',
        Low: 'priority:: low',
        Lowest: 'priority:: lowest',
        None: '',
    },
    startDateSymbol: 'start::',
    createdDateSymbol: 'created::',
    scheduledDateSymbol: 'scheduled::',
    dueDateSymbol: 'due::',
    doneDateSymbol: 'completion::',
    recurrenceSymbol: 'repeat::',
    TaskFormatRegularExpressions: {
        priorityRegex: toInlineFieldRegex(/priority:: *(highest|high|medium|low|lowest)/),
        startDateRegex: toInlineFieldRegex(/start:: *(\d{4}-\d{2}-\d{2})/),
        createdDateRegex: toInlineFieldRegex(/created:: *(\d{4}-\d{2}-\d{2})/),
        scheduledDateRegex: toInlineFieldRegex(/scheduled:: *(\d{4}-\d{2}-\d{2})/),
        dueDateRegex: toInlineFieldRegex(/due:: *(\d{4}-\d{2}-\d{2})/),
        doneDateRegex: toInlineFieldRegex(/completion:: *(\d{4}-\d{2}-\d{2})/),
        recurrenceRegex: toInlineFieldRegex(/repeat:: *([a-zA-Z0-9, !]+)/),
    },
} as const;

/**
 * A {@link TaskSerializer} that that reads and writes tasks compatible with
 *   [Dataview]{@link https://github.com/blacksmithgu/obsidian-dataview}
 */
export class DataviewTaskSerializer extends DefaultTaskSerializer {
    constructor() {
        super(DATAVIEW_SYMBOLS);
    }

    protected parsePriority(p: string): Priority {
        switch (p) {
            case 'highest':
                return Priority.Highest;
            case 'high':
                return Priority.High;
            case 'medium':
                return Priority.Medium;
            case 'low':
                return Priority.Low;
            case 'lowest':
                return Priority.Lowest;
            default:
                return Priority.None;
        }
    }

    public componentToString(task: Task, layout: TaskLayout, component: TaskLayoutComponent) {
        const stringComponent = super.componentToString(task, layout, component);
        const notInlineFieldComponents: TaskLayoutComponent[] = ['blockLink', 'description'];
        const shouldMakeInlineField = stringComponent !== '' && !notInlineFieldComponents.includes(component);
        return shouldMakeInlineField
            ? // Having 2 (TWO) leading spaces avoids a rendering issues that makes every other
              // square-bracketed inline-field invisible.
              // See https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1913
              `  [${stringComponent.trim()}]`
            : stringComponent;
    }
}
