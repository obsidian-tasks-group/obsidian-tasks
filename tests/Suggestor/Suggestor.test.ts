/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import * as chrono from 'chrono-node';
import { getSettings } from '../../src/Config/Settings';
import type { SuggestInfo } from '../../src/Suggestor';
import { makeDefaultSuggestionBuilder } from '../../src/Suggestor/Suggestor';
import { DEFAULT_SYMBOLS } from '../../src/TaskSerializer/DefaultTaskSerializer';
import { DATAVIEW_SYMBOLS } from '../../src/TaskSerializer/DataviewTaskSerializer';
import { MarkdownTable } from '../TestingTools/VerifyMarkdownTable';

window.moment = moment;

// Set predictable date for all tests in this file
const mockDate = new Date(moment('2022-07-11 15:00').valueOf());

const chronoSpy = jest
    .spyOn(chrono, 'parseDate')
    .mockImplementation((text, _, options) => chrono.en.casual.parseDate(text, mockDate, options)!);

afterAll(() => {
    chronoSpy.mockRestore();
});

const MAX_GENERIC_SUGGESTIONS_FOR_TESTS = 50;

describe.each([
    { name: 'emoji', symbols: DEFAULT_SYMBOLS },
    { name: 'dataview', symbols: DATAVIEW_SYMBOLS },
])("auto-complete with '$name' symbols", ({ symbols }) => {
    const buildSuggestions = makeDefaultSuggestionBuilder(symbols, MAX_GENERIC_SUGGESTIONS_FOR_TESTS);
    const {
        dueDateSymbol,
        scheduledDateSymbol,
        startDateSymbol,
        createdDateSymbol,
        recurrenceSymbol,
        prioritySymbols,
    } = symbols;
    it('offers basic completion options for an empty task', () => {
        // Arrange
        const originalSettings = getSettings();
        const line = '- [ ] ';
        const suggestions: SuggestInfo[] = buildSuggestions(line, 5, originalSettings);
        expect(suggestions).toEqual([
            { suggestionType: 'empty', displayText: '⏎', appendText: '\n' },
            { displayText: `${dueDateSymbol} due date`, appendText: `${dueDateSymbol} ` },
            { displayText: `${startDateSymbol} start date`, appendText: `${startDateSymbol} ` },
            { displayText: `${scheduledDateSymbol} scheduled date`, appendText: `${scheduledDateSymbol} ` },
            { displayText: `${prioritySymbols.High} high priority`, appendText: `${prioritySymbols.High} ` },
            { displayText: `${prioritySymbols.Medium} medium priority`, appendText: `${prioritySymbols.Medium} ` },
        ]);
    });

    it('offers generic due date completions', () => {
        // Arrange
        const originalSettings = getSettings();
        const line = `- [ ] some task ${dueDateSymbol}`;
        const suggestions: SuggestInfo[] = buildSuggestions(line, 17, originalSettings);
        expect(suggestions[0].displayText).toContain('today');
        expect(suggestions[1].displayText).toContain('tomorrow');
        expect(suggestions.length).toEqual(6);
    });

    it('offers specific due date completions', () => {
        // Arrange
        const originalSettings = getSettings();
        const line = `- [ ] some task ${dueDateSymbol} to`;
        const suggestions: SuggestInfo[] = buildSuggestions(line, 20, originalSettings);
        expect(suggestions[0].displayText).toContain('today');
        expect(suggestions[1].displayText).toContain('tomorrow');
    });

    it('offers generic recurrence completions', () => {
        // Arrange
        const originalSettings = getSettings();
        const line = `- [ ] some task ${recurrenceSymbol}`;
        const suggestions: SuggestInfo[] = buildSuggestions(line, 17, originalSettings);
        expect(suggestions[0].displayText).toEqual('every');
        expect(suggestions[1].displayText).toEqual('every day');
        expect(suggestions[2].displayText).toEqual('every week');
    });

    it('offers specific recurrence completions', () => {
        // Arrange
        const originalSettings = getSettings();
        const line = `- [ ] some task ${recurrenceSymbol} every w`;
        const suggestions: SuggestInfo[] = buildSuggestions(line, 25, originalSettings);
        expect(suggestions[0].displayText).toEqual('every week');
        expect(suggestions[1].displayText).toEqual('every week on Sunday');
        expect(suggestions[2].displayText).toEqual('every week on Monday');
    });

    it('respects the minimal match setting', () => {
        // Arrange
        const originalSettings = getSettings();
        originalSettings.autoSuggestMinMatch = 2;
        let line = `- [ ] some task ${recurrenceSymbol} e`;
        let suggestions: SuggestInfo[] = buildSuggestions(line, 19, originalSettings);
        expect(suggestions.length).toEqual(0);
        line = `- [ ] some task ${recurrenceSymbol} ev`;
        suggestions = buildSuggestions(line, 20, originalSettings);
        expect(suggestions[0].displayText).toEqual('every');
        expect(suggestions[1].displayText).toEqual('every day');
    });

    it('matches created property suggestion when user types "created" but not "today"', () => {
        // Arrange
        const originalSettings = getSettings();
        let line = '- [ ] some task cr';
        let suggestions: SuggestInfo[] = buildSuggestions(line, 18, originalSettings);
        expect(suggestions[0].displayText).toEqual(`${createdDateSymbol} created today (2022-07-11)`);

        line = '- [ ] some task tod';
        suggestions = buildSuggestions(line, 19, originalSettings);
        expect(suggestions[0].suggestionType).toEqual('empty');
        expect(suggestions[0].displayText).not.toContain('created today');
    });

    it('show all suggested text', () => {
        const originalSettings = getSettings();
        originalSettings.autoSuggestMaxItems = 200;

        const lines = [
            '- [ ] some task',
            `- [ ] some task ${recurrenceSymbol} `,
            `- [ ] some task ${dueDateSymbol} `,
            `- [ ] some task ${scheduledDateSymbol} `,
            `- [ ] some task ${startDateSymbol} `,
        ];
        const markdownTable = new MarkdownTable(['suggestion', 'expansion']);
        for (const line of lines) {
            const suggestions: SuggestInfo[] = buildSuggestions(line, line.length - 1, originalSettings);
            for (const suggestion of suggestions) {
                // The 'new line' replacement adds a trailing space at the end of a line,
                // which causes auto-formatting to then make the test fail.
                // So we replace the 'new line' character with some fixed text,
                // that displays nicely when the output Markdown table is viewed.
                let replacementText = `${suggestion.appendText}`;
                if (replacementText === '\n') {
                    replacementText = '&lt;new line>';
                }
                markdownTable.addRowIfNew([suggestion.displayText, replacementText]);
            }
        }

        // For help if this test fails and you are new to Approval Tests, see:
        //    https://publish.obsidian.md/tasks-contributing/Testing/Approval+Tests
        markdownTable.verify();
    });
});
