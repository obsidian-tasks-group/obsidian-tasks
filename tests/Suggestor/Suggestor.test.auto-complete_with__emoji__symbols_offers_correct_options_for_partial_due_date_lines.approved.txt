How due date suggestions are affected by what the user has typed:


For this markdown line:
"- [ ] some task ðŸ“…"

The first suggestion is:
{
    "suggestionType": "match",
    "displayText": "today (2022-07-11)",
    "appendText": "ðŸ“… 2022-07-11 ",
    "insertAt": 16,
    "insertSkip": 2
}
--------------------------------------------------------------------------------

For this markdown line:
"- [ ] some task ðŸ“… 27 oct"

The first suggestion is:
{
    "suggestionType": "match",
    "displayText": "2022-10-27",
    "appendText": "ðŸ“… 2022-10-27 ",
    "insertAt": 16,
    "insertSkip": 9
}
--------------------------------------------------------------------------------

For this markdown line:
"- [ ] some task ðŸ“… 1 year"

The first suggestion is:
{
    "suggestionType": "match",
    "displayText": "2023-07-11",
    "appendText": "ðŸ“… 2023-07-11 ",
    "insertAt": 16,
    "insertSkip": 9
}
--------------------------------------------------------------------------------

